---
title: "RISE_application_ver2"
author: "Arthur Hughes"
date: "2025-01-15"
output: html_document
editor_options: 
  chunk_output_type: console
---

# Setup
Load packages, functions, and data. 
```{r, eval = F}
# Packages
library(dplyr)
library(ggplot2)
library(tidyr)
library(stringr)
library(SurrogateRank)
library(readr)
library(parallel)
library(pbmcapply)
library(caTools)
library(here)
library(xtable)
library(cowplot)
library(caret)
library(ggrepel)
library(GSA)

# Load functions for RISE
# Define the folder path
folder_path <- "./R/"
# Get all .R files in the folder
r_files <- list.files(path = folder_path, pattern = "\\.R$", full.names = TRUE)
# Source each file
sapply(r_files, source)

myxtable <- function(x, ...) xtable(apply(x, 2, as.character), ...) # Function to convert dataframe output into latex table output

# Merged data containing characteristics, immune responses, and gene expression
hipc_merged = get(load("~/Bureau/Work/PhD/data management/data/processed/hipc_merged.RData"))
```

Subset the data by female recipients of the 2008 inactivated influenza vaccines.
```{r}
set.seed(18042025)
split_percent = 0.75

# 2008 Inactivated influenza data
hipc_flu = hipc_merged %>% 
  filter(pathogen == "Influenza",
         vaccine_type == "Inactivated",
         vaccine == "TIV (2008)",
         gender == "Female",
         time_post_last_vax %in% c(0,1)) %>% # gene expression at day 0 or day 1
  mutate(
    treatment = ifelse(time_post_last_vax == 0, 0, 1), # set treatment to 0 for baseline gene expression samples and 1 for post-vax gene expression
    response = ifelse(time_post_last_vax == 0, # set response to day 0 nAb for baseline gene expression samples and day 28 nAb for post-vax gene expression
                      n_ab_mean_value_d0,
                      n_ab_mean_value_d28)) %>%
  filter(!is.na(response)) %>%  # remove samples without a response 
  dplyr::select(participant_id, treatment, response, c(a1cf:zzz3)) # select treatment, repsonse, all gene expression
```

# Preprocessing data
Put the data in paired format
```{r}
# Total number of genes in the data
n_genes = ncol(hipc_flu %>% dplyr::select(a1cf:zzz3))

gene_names = hipc_flu %>% dplyr::select(a1cf:zzz3) %>% colnames()

# Gene names
all_predictors <- c("response", gene_names)

# Sort data by individual, remove unpaired data
paired_long_df <- hipc_flu %>%
  dplyr::select(participant_id, treatment, all_of(all_predictors)) %>%
  group_by(participant_id) %>%
  filter(n_distinct(treatment) == 2) %>%  # keep only individuals with both treatments
  ungroup() %>%
  arrange(participant_id, treatment)

# Sample splitting
train_ids <- paired_long_df %>%
  distinct(participant_id) %>%
  slice_sample(prop = split_percent) %>%
  pull(participant_id)

# Step 2: Filter the long dataframe to include only the sampled participants
hipc_flu_train <- paired_long_df %>%
  filter(participant_id %in% train_ids)

hipc_flu_test <- paired_long_df %>%
  filter(!(participant_id %in% train_ids))
```


# Screening Stage
Apply the statistical test one-by-one to each gene in the data to identify a group of strong candidate surrogates. 

```{r}
X = hipc_flu_train %>% dplyr::select(-c(participant_id, response, treatment)) %>% as.matrix() # Candidate gene expression surrogates
Y = hipc_flu_train %>% dplyr::select(response) %>% as.matrix() # Response
A = hipc_flu_train %>% dplyr::select(treatment) %>% as.matrix() # Treatment groups

uy_check = test.surrogate(yone = Y[A==1],
                                 yzero = Y[A==0],
                                 sone = Y[A==1],
                                 szero = Y[A==0],
                                 mode = "paired",
                                 test = "non.inferiority")$u.y

uy_check

# yone = Y[A==1]
# yzero = Y[A==0]
# sone = X[A==1]
# szero = X[A==0]
# epsilon = NULL
# full.data = NULL
# power.want.s = 0.9
# u.y.hyp = NULL
# alpha = 0.05
# test = "two.one.sided"
# mode = "paired"


# Screen genes in parallel
## Desired power of 90% for new test of treatment effect based on surrogates
## We reject if bonferroni adjusted p is below alpha = 0.05
res_screen = rise_screen(Y = Y, X = X, A = A,
                         reference = "0", alpha = 0.05,
                         power_desired = 0.9, p_correction = "bonferroni",
                         cores = 9,mode = "paired",
                         test = "two.one.sided")

# Save the results as a dataframe
res_df = res_screen$results

res_screen$significant_markers %>% length()
```

## Table 1
Screening results from the data application - top 20 genes by adjusted p-values.
```{r, eval = F}
# Can paste directly into latex table
latex_results = res_screen$results %>% 
  arrange(p_unadjusted) %>% 
  filter(p_adjusted < 0.05) %>% 
  mutate(marker = toupper(marker),
         delta = round(delta, 3),
         p_unadjusted = formatC(p_unadjusted, format = "e", digits = 1),
         p_adjusted = formatC(p_adjusted, format = "e", digits = 1),
         delta = paste0(round(delta, 3), " (", round(ci_lower, 3), ", ", round(ci_upper, 3), ")"),
         sd = round(sd, 3)) %>% 
  dplyr::select(marker, delta, sd, p_unadjusted, p_adjusted)

# print latex table
print(myxtable(latex_results %>% head(20)), include.rownames = FALSE)
```

## Table 2
Functional annotation of 235 significant genes using DAVID. The table shows functional terms, the number ofassociated genes, and adjusted p-values for a test of term overrepresentation compared to a random sample of the same size fromthe 10,086 aligned genes.
```{r, eval = F}
# Output string to paste into DAVID function gene enrichment tool
david = cat(toupper(res_screen$significant_markers),sep = ", ")

# Background list for DAVID
david_background = cat(toupper(gene_names),sep = ", ")

# Download the chart as a .txt file from DAVID. Load the data in and output it at a latex table. 
david_df = read.table("./output/application/david.txt", header = TRUE, sep = "\t") 

david_df = david_df %>% 
  dplyr::select(Term, Count, Benjamini) %>% 
  mutate(Benjamini = formatC(Benjamini, format = "e", digits = 1),
         Term = gsub("~", "-", Term)
  )

print(myxtable(david_df %>% head(10)), include.rownames = FALSE)
```


# Evaluation

## Figure 5
Figure 5 : Ranks of the mean neutralising antibodies against the ranks of the constructed 235-gene-combination surrogate marker in the evaluation dataset. The Spearman rank correlation coefficient is 0.79, indicating strong positive correlation.
```{r}
# Select only the significant genes from screening
genes = res_screen$significant_markers
# genes = (res_df %>% arrange(p_unadjusted) %>% head(10))$marker
# train_index = get(load(here("output","application","train_index_final.RData")))
hipc_flu_test <- paired_long_df %>%
  filter(!(participant_id %in% train_ids))

X_evaluate = hipc_flu_test %>% dplyr::select(all_of(genes)) %>% as.matrix() # Candidate gene expression surrogates
Y_evaluate = hipc_flu_test %>% dplyr::select(response) %>% as.matrix() # Response
A_evaluate = hipc_flu_test %>% dplyr::select(treatment) %>% as.matrix() # Treatment

# Check treatment effect
test.surrogate(yone = Y_evaluate[A_evaluate==1],
                      yzero = Y_evaluate[A_evaluate==0],
                      sone = Y_evaluate[A_evaluate==1],
                      szero = Y_evaluate[A_evaluate==0], 
                      mode = "paired", test = "two.one.sided")$u.y


weights = res_screen[["weights"]]  # use the absolute value of the weights from the screening stage
weights[,2] = abs(weights[,2])
evaluate_results = rise_evaluate(Y_evaluate = Y_evaluate,
                                 X_evaluate = X_evaluate,
                                 A_evaluate = A_evaluate,
                                 individual = T,
                                 power_desired = 0.9,
                                 plot = T,
                                 markers = genes, 
                                 weights = weights,
                                 mode = "paired", test = "two.one.sided") # results for the combined predictor 

# plot of ranks of primary response vs ranks of combined predictor
fig5 <- evaluate_results$plot +
  geom_point(size = 8) +
  xlab("Neutralising Antibodies (rank)") +
  ylab(expression(gamma[S]~"(rank)")) +
  ggtitle("Ranks of primary outcome vs combined surrogate in evaluation data") +
  guides(
    col = guide_legend(
      title = "Sample",
      override.aes = list(size = 10) # Increase point size in legend
    )
  ) +
  scale_color_manual(
    labels = c("Pre-vaccine", "Post-vaccine"), 
    values = c("#7BC4F4", "#001B52")
  ) +
  theme_minimal(base_size = 35) +
  theme(
    plot.title = element_blank(),
    axis.title.x = element_text(size = 35),
    axis.title.y = element_text(size = 45),
    legend.text = element_text(size = 25),  # Increase legend text size
    legend.title = element_text(size = 35), # Increase legend title size
    plot.background = element_rect(fill = 'white', color = 'white')
  )

fig5

ggsave(plot = fig5, path = here("output", "figures"),
       filename = "figure5.eps",
       units = "cm", width = 35, height = 25, dpi = 800)
# 
ggsave(plot = fig5, path = here("output", "figures"),
       filename = "figure5.pdf",
       units = "cm", width = 35, height = 25, dpi = 800)


evaluate_results$evaluation_results
```

## Where do the individuals go?

```{r}
gamma_evaluate = evaluate_results$gamma

# Creating an example data frame
data_paired_plot <- data.frame(
  gamma_S = gamma_evaluate, 
  A_evaluate = factor(A_evaluate), 
  Y_evaluate = Y_evaluate,
  individual_id = rep(1:(length(gamma_evaluate) / 2), each = 2)  # Assuming paired data
)

# Plotting using ggplot2
ggplot(data_paired_plot, aes(x = Y_evaluate, y = gamma_S, color = A_evaluate, group = individual_id)) +
  geom_line(aes(group = individual_id), color = "gray", alpha = 0.5) +  # Trajectories
  geom_point(size = 5) +  # Points for each (Y_evaluate, gamma) pair
  geom_smooth(method = "lm", se = FALSE, aes(group = 1), color = "red", linetype = "dashed") +  # Regression line
  labs(x = "Y_evaluate", y = "Gamma", color = "A_evaluate") +
  theme_minimal() +
  xlab("Primary Response") +
  ylab(expression(gamma[S]~"")) +
  guides(
    col = guide_legend(
      title = "Sample",
      override.aes = list(size = 10) # Increase point size in legend
    )
  ) +
  scale_color_manual(
    labels = c("Pre-vaccine", "Post-vaccine"), 
    values = c("#7BC4F4", "#001B52")
  ) + 
  ggtitle("Ranks of primary response vs new surrogate in evaluation data with individual trajectories")+
  annotate(size = 15, 
           "text",
           x = Inf, 
           y = Inf, 
           label = bquote(rho == .(round(cor(as.numeric(data_paired_plot[,1]), as.numeric(data_paired_plot[,3]), method = "pearson"), 2))), 
           vjust = 2, 
           hjust = 3.3, 
           color = "red") + 
  theme_minimal(base_size = 22) +
  theme(plot.title = element_text(hjust = 0.5, face = "bold"),
        plot.background = element_rect(fill = 'white', color = 'white'))

ggplot(data_paired_plot, aes(x = rank(Y_evaluate), y = rank(gamma_S), color = A_evaluate, group = individual_id)) +
  geom_line(aes(group = individual_id), color = "gray", alpha = 0.5) +  # Trajectories
  geom_point(size = 5) +  # Points for each (Y_evaluate, gamma) pair
  labs(x = "Y_evaluate", y = "Gamma", color = "A_evaluate") +
  theme_minimal() +
  xlab("Primary Response (rank)") +
  ylab(expression(gamma[S]~"(rank)")) +
  guides(
    col = guide_legend(
      title = "Sample",
      override.aes = list(size = 10) # Increase point size in legend
    )
  ) +
  scale_color_manual(
    labels = c("Pre-vaccine", "Post-vaccine"), 
    values = c("#7BC4F4", "#001B52")
  ) + 
  ggtitle("Ranks of primary response vs new surrogate in evaluation data with individual trajectories")+
  coord_fixed(ratio = 1) +
  geom_abline(slope=1, col= "red", linetype = "dashed", size = 1.4) +
  annotate(size = 15, 
           "text",
           x = Inf, 
           y = Inf, 
           label = bquote(rho == .(round(cor(as.numeric(data_paired_plot[,1]), as.numeric(data_paired_plot[,3]), method = "spearman"), 2))), 
           vjust = 2, 
           hjust = 3.3, 
           color = "red") + 
  theme_minimal(base_size = 22) +
  theme(plot.title = element_text(hjust = 0.5, face = "bold"),
        plot.background = element_rect(fill = 'white', color = 'white'))


# Change in Y vs change in gamma plot
data_diff = data_paired_plot %>%
  # Group by individual_id to calculate changes within each individual
  group_by(individual_id) %>%
  # Spread the data so we can calculate the change between A_evaluate = 1 and A_evaluate = 0
  summarise(
    change_gamma = gamma_S[A_evaluate == 1] - gamma_S[A_evaluate == 0],
    change_response = response[A_evaluate == 1] - response[A_evaluate == 0]
  ) %>%
  # Ungroup to return to normal data
  ungroup()

# Plot the differences
ggplot(data_diff, aes(x = change_response, y = change_gamma)) +
  geom_point(size = 5) +  # Points for each individual
  labs(x = "Difference in Y_evaluate", y = "Difference in Gamma") +
  theme_minimal(base_size = 22) +
  # coord_fixed(ratio = 1) +
  # geom_abline(slope=1, col= "red", linetype = "dashed", size = 1.4) +
  geom_smooth(method = "lm", color = "red", linetype = "dashed", size = 1.4,se = F) 

ggplot(data_diff, aes(x = rank(change_response), y = rank(change_gamma))) +
  geom_point(size = 5) +  # Points for each individual
  labs(x = "Difference in Y_evaluate (rank)", y = "Difference in Gamma (rank)") +
  theme_minimal(base_size = 22) +
  coord_fixed(ratio = 1) +
  geom_abline(slope=1, col= "red", linetype = "dashed", size = 1.4) +
  annotate(size = 15, 
           "text",
           x = Inf, 
           y = Inf, 
           label = bquote(rho == .(round(cor(data_diff[,2], data_diff[,3], method = "spearman"), 2))), 
           vjust = 2, 
           hjust = 3.3, 
           color = "red")

```




# Supporting Information
## Table S1
Supplementary Table 1: Screening results from the data application - all genes with adjusted p-values less than 0.05.
```{r, eval = F}
# Can paste directly into latex table
latex_results = res_screen$results %>% 
  arrange(p_unadjusted) %>% 
  filter(p_adjusted < 0.05) %>% 
  mutate(marker = toupper(marker),
         delta = round(delta, 3),
         p_unadjusted = formatC(p_unadjusted, format = "e", digits = 1),
         p_adjusted = formatC(p_adjusted, format = "e", digits = 1),
         delta = paste0(round(delta, 3), " (", round(ci_lower, 3), ", ", round(ci_upper, 3), ")"),
         sd = round(sd, 3)) %>% 
  dplyr::select(marker, delta, sd, p_unadjusted, p_adjusted)

# print latex table
print(myxtable(latex_results), include.rownames = FALSE)
```

## Table S2
Supplementary Table S2: Sensitivity analysis evaluating the effect of varying the non-inferiority margin $\epsilon$, where values closer to 0 result in fewer candidate surrogates to combine for the evaluation stage. The evaluation metric for the combined marker, $\boldsymbol{\delta_{\gamma_{S}}}$, its standard deviation, and its p-value corresponding to a test based on a desired power of 90\% are given in the table.
```{r, eval = F}
X = hipc_flu_train %>% dplyr::select(-c(participant_id, response, treatment)) %>% as.matrix() # Candidate gene expression surrogates
Y = hipc_flu_train %>% dplyr::select(response) %>% as.matrix() # Response
A = hipc_flu_train %>% dplyr::select(treatment) %>% as.matrix() # Treatment groups
epsilon_grid = c(0.05, 0.1, 0.15, 0.2, 0.25, 0.3, 0.35)
res_screen_sensitivity = list()

for (i in 1:length(epsilon_grid)){
  res_screen_sensitivity[[i]] = rise_screen(
    Y = Y, 
    X = X, 
    A = A,
    reference = "0", 
    alpha = 0.05,
    epsilon = epsilon_grid[i], 
    p_correction = "bonferroni",
    cores = 9,
    mode = "paired",
    test = "two.one.sided"
  )
}

# save the results
save(res_screen_sensitivity, file = './output/RISE_sensitivity_analysis.Rds')
res_screen_sensitivity = get(load('./output/RISE_sensitivity_analysis.Rds'))

significant_genes = list()
n_significant_genes = rep(0,length(epsilon_grid))
for (i in 1:length(epsilon_grid)){
  significant_genes[[i]] = res_screen_sensitivity[[i]]$significant_markers
  n_significant_genes[i] = length(significant_genes[[i]])
}


# evaluation
source("./R/rise_evaluate.R")
evaluate_results_sensitivity = list()
for (i in 1:length(epsilon_grid)){
  
    if(n_significant_genes[i] == 0){
    next
  }
  
  markers = significant_genes[[i]]
  X_evaluate = hipc_flu_test %>% dplyr::select(all_of(markers)) %>% as.matrix()
  Y_evaluate = hipc_flu_test %>% dplyr::select(response) %>% as.matrix()
  A_evaluate = hipc_flu_test %>% dplyr::select(treatment) %>% as.matrix()
  weights = res_screen_sensitivity[[i]][["weights"]]
  weights[,2] = abs(weights[,2])
  evaluate_results_sensitivity[[i]] = rise_evaluate(
    Y_evaluate = Y_evaluate,
    X_evaluate = X_evaluate,
    A_evaluate = A_evaluate,
    individual = T,
    power_desired = 0.9,
    plot = T,
    markers = markers, 
    weights = weights,
    mode = "paired", 
    test = "two.one.sided")
  print(evaluate_results_sensitivity[[i]][["evaluation_results"]])
  print(beta_grid[i])
  print(evaluate_results_sensitivity[[i]]$plot)
  
}

# Identify the expected column names (and length) from a non-empty example
output = bind_cols("epsilon_screen" = epsilon_grid[-c(1,2)], "n_predictors" =  n_significant_genes[-c(1,2)], bind_rows(lapply(evaluate_results_sensitivity, function(x) x$evaluation_results))) %>% 
  as.data.frame() %>% 
  mutate(delta_ci = paste0(round(delta, 3), " (", round(ci_lower, 3), ", ", round(ci_upper, 3), ")"),
         sd_round = round(sd, 3),
         p_unadjusted = formatC(p_unadjusted, format = "e", digits = 1)) %>% 
  dplyr::select(-marker)

results_sensitivity_latex = output %>% dplyr::select(epsilon_screen, n_predictors, delta_ci, sd_round, p_unadjusted)
myxtable <- function(x, ...) xtable(apply(x, 2, as.character), ...)
print(myxtable(results_sensitivity_latex), include.rownames = FALSE)
```

## Figure S8
Supplementary Figure 8: A visual method to select markers to pass the screening stage. The x-axis is the negative log of the $\delta$ values, where $0.05$ is added to avoid taking the log of 0, and the y-axis is the negative log of the raw p-value. Markers with a stronger surrogate strength appear towards the top-right of the plot. One could select markers without relying on a p-value threshold by choosing those which are most separated from the rest. In this example, the top 16 genes are labelled as they visually separate themselves. Evaluating the combined surrogacy of these genes gives $\delta 95\%\text{C.I.} = -0.01 (-1,0.009)$ corresponding to $p = 2.1e-40$. Points in red are the 235 genes which were selected according to the adjusted p-value threshold of 0.05. 
```{r}
res_df_filtered = res_df %>% 
  filter(delta < 1)

res_df_filtered$marker = res_df_filtered$marker %>% toupper()
# Replace 0s with a small value to avoid -log10(0) = Inf
res_df_filtered$p_adjusted[res_df_filtered$p_adjusted == 0] <- 1e-300

# Add significance columns
res_df_filtered$significant <- res_df_filtered$p_adjusted < 0.05
res_df_filtered$p_group <- ifelse(res_df_filtered$significant, "<0.05", ">0.05")

# Select top 10 most significant genes based on smallest p_adjusted
top10 <- res_df_filtered[order(res_df_filtered$p_adjusted), ][1:17, ]

# Make sure the selected rows have valid values for delta and p_adjusted
top10 <- top10[is.finite(top10$delta) & is.finite(top10$p_adjusted), ]

# Create the plot with arrows and labels
figs8 = ggplot(res_df_filtered, aes(x = delta, y = -log10(p_adjusted), color = p_group)) +
  geom_point(size = 2, alpha = 0.2) +  # Reduced point size for better visibility
  geom_text_repel(
    data = top10,
    aes(label = marker),  # Assuming 'marker' column contains the gene names or identifiers
    color = "red",
    size = 9,  # Adjust label size
    arrow = arrow(type = "closed", length = unit(0, "inches")),  # Adjust arrow size
    box.padding = 0.5,  # Increased padding to separate labels from points
    nudge_x = 0.3,  # Nudging the label horizontally to avoid overlap
    nudge_y = 0.5,  # Nudging the label vertically to avoid overlap
    max.overlaps = 10,  # Limit the number of overlaps
    segment.color = "red",  # Color of the arrow
    segment.alpha = 0.2  # Adjust transparency of the arrow
  ) +
  scale_color_manual(
    values = c("<0.05" = "red", ">0.05" = "black"),
    labels = c("<0.05", ">0.05")
  ) +
  labs(
    x = expression(delta),  # LaTeX-style Î´
    y = expression(-log[10](p[adj])),
    color = expression(p[adj])
  ) +
  xlim(-0.1,1) +
  theme_minimal(base_size = 30) +
  theme()  # Remove text from the legend

# figs8

ggsave(plot = figs8,
       here("output","figures","figureS8.pdf"), 
       units = "cm", width = 40, height = 25)
```


## Debugging

```{r}
full.data = NULL
epsilon = NULL
power.want.s = 0.8
u.y.hyp = NULL
alpha = 0.05 
mode = "paired" 
test = "two.one.sided"

gamma_evaluate = evaluate_results$gamma
Y_evaluate = hipc_flu_test %>% dplyr::select(response) %>% as.matrix() # Response
A_evaluate = hipc_flu_test %>% dplyr::select(treatment) %>% as.matrix() #


yzero = Y_evaluate[A_evaluate == 0]
yone = Y_evaluate[A_evaluate == 1]

szero = gamma_evaluate[A_evaluate == 0]
sone = gamma_evaluate[A_evaluate == 1]


```